File: packages/wire/padding_test.go
package wire

import (
	"bytes"
	"crypto/rand"
	"testing"
)

func TestPadBucketSize(t *testing.T) {
	tests := []struct {
		unpaddedLen int
		wantBucket  int
	}{
		{0, 256},
		{1, 256},
		{100, 256},
		{254, 256}, // 254 + 2 = 256
		{255, 512}, // 255 + 2 = 257 -> 512
		{510, 512}, // 510 + 2 = 512
		{511, 1024},
		{1000, 1024},
		{1022, 1024},
		{1023, 2048},
		{4000, 4096},
		{4094, 4096},
		{4095, 8192},
		{16382, 16384},
		{16383, 32768},
		{32766, 32768},
		{32767, 65535},
		{60000, 65535},
		{65533, 65535},
	}

	for _, tc := range tests {
		got := PadBucketSize(tc.unpaddedLen)
		if got != tc.wantBucket {
			t.Errorf("PadBucketSize(%d) = %d, want %d", tc.unpaddedLen, got, tc.wantBucket)
		}
	}
}

func TestPadAndUnpadPayload(t *testing.T) {
	lengths := []int{0, 1, 10, 254, 255, 500, 1022, 1023, 4094, 16382, 32766, 65533}
	for _, l := range lengths {
		payload := make([]byte, l)
		if l > 0 {
			_, _ = rand.Read(payload)
		}

		padded, err := PadPayload(payload)
		if err != nil {
			t.Fatalf("PadPayload(len=%d): %v", l, err)
		}

		expectedBucket := PadBucketSize(l)
		if len(padded) != expectedBucket {
			t.Errorf("len(padded) = %d, want bucket %d", len(padded), expectedBucket)
		}

		unpadded, err := UnpadPayload(padded)
		if err != nil {
			t.Fatalf("UnpadPayload(len=%d): %v", l, err)
		}

		if !bytes.Equal(unpadded, payload) {
			t.Errorf("unpadded != original payload for len %d", l)
		}
	}
}

func TestUnpadPayloadFailClosed(t *testing.T) {
	t.Run("truncated buffer", func(t *testing.T) {
		_, err := UnpadPayload([]byte{0x00})
		if err == nil {
			t.Error("expected error for 1-byte buffer")
		}
	})

	t.Run("length field exceeds buffer", func(t *testing.T) {
		buf := make([]byte, 256)
		buf[0] = 0x01
		buf[1] = 0x00 // declared length 256, but buffer is only 256 (needs 256 + 2 = 258)
		_, err := UnpadPayload(buf)
		if err == nil {
			t.Error("expected error when length + 2 > buffer")
		}
	})

	t.Run("non-zero padding byte", func(t *testing.T) {
		buf := make([]byte, 256)
		buf[0] = 0x00
		buf[1] = 0x04 // declared length 4
		copy(buf[2:6], []byte("test"))
		buf[100] = 0x01 // tampered padding byte
		_, err := UnpadPayload(buf)
		if err == nil {
			t.Error("expected error when padding contains non-zero bytes")
		}
	})
}

func TestSealPaddedAndOpen(t *testing.T) {
	key := make([]byte, aeadKeyBytes)
	salt := make([]byte, aeadSaltBytes)
	_, _ = rand.Read(key)
	_, _ = rand.Read(salt)
	dir := DirectionalKey{Key: key, Salt: salt}

	h := FrameHeaderInput{
		Version:  FrameVersion,
		Type:     FrameBlockData,
		Flags:    FrameFlagLastInBlock,
		FileIdx:  0,
		BlockIdx: 1,
		FrameOff: 0,
	}

	payload := []byte("Hello, private padded world!")
	sealed, err := SealPadded(dir, 1, h, payload)
	if err != nil {
		t.Fatalf("SealPadded: %v", err)
	}

	opened, err := Open(dir, 1, sealed)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}

	if opened.Header.Flags&FrameFlagPadded == 0 {
		t.Error("expected FrameFlagPadded in opened header")
	}
	if opened.Header.Flags&FrameFlagLastInBlock == 0 {
		t.Error("expected FrameFlagLastInBlock preserved in opened header")
	}
	if !bytes.Equal(opened.Plaintext, payload) {
		t.Errorf("got plaintext %q, want %q", opened.Plaintext, payload)
	}
}

File: packages/protocol/src/padding.test.ts
import { describe, it, expect } from 'vitest';
import { FRAME_FLAG_LAST_IN_BLOCK, FRAME_FLAG_PADDED, FRAME_VERSION } from './constants.js';
import { padBucketSize, padPayload, unpadPayload } from './frame.js';
import { open, openSequenced, sealPadded } from './aead.js';
import { FrameType } from './transfer.js';
import { deriveTransferKeys } from './keyschedule.js';
import { sha256 } from './webcrypto.js';
import { MemorySink, bytesSource } from './transfer-ports.js';
import { TransferSender } from './transfer-sender.js';
import { TransferReceiver } from './transfer-receiver.js';

describe('traffic padding', () => {
  it('computes correct power-of-two bucket sizes', () => {
    expect(padBucketSize(0)).toBe(256);
    expect(padBucketSize(1)).toBe(256);
    expect(padBucketSize(100)).toBe(256);
    expect(padBucketSize(254)).toBe(256);
    expect(padBucketSize(255)).toBe(512);
    expect(padBucketSize(510)).toBe(512);
    expect(padBucketSize(511)).toBe(1024);
    expect(padBucketSize(1000)).toBe(1024);
    expect(padBucketSize(1022)).toBe(1024);
    expect(padBucketSize(1023)).toBe(2048);
    expect(padBucketSize(4000)).toBe(4096);
    expect(padBucketSize(4094)).toBe(4096);
    expect(padBucketSize(4095)).toBe(8192);
    expect(padBucketSize(16382)).toBe(16384);
    expect(padBucketSize(16383)).toBe(32768);
    expect(padBucketSize(32766)).toBe(32768);
    expect(padBucketSize(32767)).toBe(65535);
    expect(padBucketSize(60000)).toBe(65535);
    expect(padBucketSize(65533)).toBe(65535);
  });

  it('pads and unpads payloads cleanly across multiple lengths', () => {
    const lengths = [0, 1, 10, 254, 255, 500, 1022, 1023, 4094, 16382, 32766, 65533];
    for (const len of lengths) {
      const payload = new Uint8Array(len);
      for (let i = 0; i < len; i++) {
        payload[i] = (i * 11 + 7) % 256;
      }

      const padded = padPayload(payload);
      expect(padded.length).toBe(padBucketSize(len));

      const unpadded = unpadPayload(padded);
      expect(unpadded).toEqual(payload);
    }
  });

  it('fails closed on malformed padding', () => {
    // 1. Truncated
    expect(() => unpadPayload(new Uint8Array([0]))).toThrow(/too short/);

    // 2. Length field exceeds buffer
    const badLen = new Uint8Array(256);
    new DataView(badLen.buffer).setUint16(0, 256, false); // needs 258 bytes
    expect(() => unpadPayload(badLen)).toThrow(/exceeds buffer/);

    // 3. Non-zero padding byte
    const badPad = new Uint8Array(256);
    new DataView(badPad.buffer).setUint16(0, 4, false);
    badPad.set(new Uint8Array([1, 2, 3, 4]), 2);
    badPad[100] = 0x01; // corrupted padding byte
    expect(() => unpadPayload(badPad)).toThrow(/non-zero padding byte/);
  });

  it('seals padded frames and recovers plaintext with open', async () => {
    const master = new Uint8Array(32).fill(0x42);
    const keys = await deriveTransferKeys(master);

    const payload = new TextEncoder().encode('Hello padded authenticated world!');
    const header = {
      version: FRAME_VERSION,
      type: FrameType.BlockData,
      flags: FRAME_FLAG_LAST_IN_BLOCK,
      fileIdx: 0,
      blockIdx: 1,
      frameOff: 0,
    };

    const sealed = await sealPadded(keys.o2j, 1, header, payload);
    const opened = await open(keys.o2j, 1, sealed);

    expect(opened.header.flags & FRAME_FLAG_PADDED).not.toBe(0);
    expect(opened.header.flags & FRAME_FLAG_LAST_IN_BLOCK).not.toBe(0);
    expect(opened.plaintext).toEqual(payload);

    // openSequenced also works
    const seqOpened = await openSequenced(keys.o2j, 1, sealed);
    expect(seqOpened.plaintext).toEqual(payload);
  });

  it('runs complete padded transfer between TransferSender and TransferReceiver', async () => {
    const master = new Uint8Array(32).fill(0x55);
    const keys = await deriveTransferKeys(master);

    const data = new Uint8Array(100 * 1024); // 100 KB
    for (let i = 0; i < data.length; i++) data[i] = (i * 13 + 17) % 256;

    const source = bytesSource(data, {
      name: 'padded-test.bin',
      size: data.length,
      mime: 'application/octet-stream',
      lastModified: 0,
    });
    const sink = new MemorySink();

    const networkQueue: Uint8Array[] = [];

    const sender = new TransferSender({
      file: source,
      send: (frame) => {
        networkQueue.push(frame);
        void deliverReceiver();
      },
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: () => {
        let buf = new Uint8Array(0);
        return {
          update: (b) => {
            const next = new Uint8Array(buf.length + b.length);
            next.set(buf);
            next.set(b, buf.length);
            buf = next;
          },
          hexDigest: async () => {
            const hash = await sha256(buf);
            return Array.from(hash)
              .map((b) => b.toString(16).padStart(2, '0'))
              .join('');
          },
        };
      },
      padding: true,
    });

    const receiver = new TransferReceiver({
      sink,
      send: (frame) => {
        void sender.handle(frame);
      },
      sendDir: keys.j2o,
      recvDir: keys.o2j,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: () => {
        let buf = new Uint8Array(0);
        return {
          update: (b) => {
            const next = new Uint8Array(buf.length + b.length);
            next.set(buf);
            next.set(b, buf.length);
            buf = next;
          },
          hexDigest: async () => {
            const hash = await sha256(buf);
            return Array.from(hash)
              .map((b) => b.toString(16).padStart(2, '0'))
              .join('');
          },
        };
      },
      padding: true,
    });

    let delivering = false;
    async function deliverReceiver() {
      if (delivering) return;
      delivering = true;
      try {
        while (networkQueue.length > 0) {
          const frame = networkQueue.shift()!;
          await receiver.handle(frame);
        }
      } finally {
        delivering = false;
      }
    }

    const [sendDigest, recvOutcome] = await Promise.all([sender.run(), receiver.done]);

    expect(sendDigest).toBe(recvOutcome.digest);
    expect(sink.bytes()).toEqual(data);
  });
});
